pkg/virt-controller/watch/replicaset.go

```diff
@@ -469,7 +487,7 @@ func (c *VMIReplicaSet) updateVirtualMachine(old, cur interface{}) {
 	}
 
 	labelChanged := !reflect.DeepEqual(curVMI.Labels, oldVMI.Labels)
-	if curVMI.DeletionTimestamp != nil {
+	if curVMI.DeletionTimestamp != nil || curVMI.IsFinal() {
```

Comment:
That looks wrong. A vm which is final is not a delete.

Reply:
@rmohr Without it, the code does not call to `c.deleteVirtualMachine` that update expectations, in this case, `needSync` variable will have `false` value and the code will skip clean and scale stages.

Reply:
If that's the case, then there is something else wrong. A deletion expectation for a UID is only fulfilled if you either see a real delete (object immediately disappears from the cluster, in this case `deleteVirtualMachine` is directly invoked by the informer) or if you get a chache update that the deletion timestamp was set (that's the situation in this line).